web/src/stream_popover.js

```diff
@@ -362,6 +363,79 @@ export function build_move_topic_to_stream_popover(current_stream_id, topic_name
     });
 }
 
+export function build_topic_edit_popover(current_stream_id, topic_name) {
+    const args = {
+        current_stream_id,
+        topic_name,
+        notify_new_thread: message_edit.notify_new_thread_default,
+        notify_old_thread: message_edit.notify_old_thread_default,
+    };
+    const current_stream_name = stream_data.maybe_get_stream_name(current_stream_id);
+
+    function rename_topic() {
+        const params = get_params_from_form($("#rename_topic_form"));
+
+        const {old_topic_name, new_topic_name} = params;
+
+        let {current_stream_id, send_notification_to_new_thread, send_notification_to_old_thread} =
+            params;
+        send_notification_to_new_thread = send_notification_to_new_thread === "on";
+        send_notification_to_old_thread = send_notification_to_old_thread === "on";
+        current_stream_id = Number.parseInt(current_stream_id, 10);
+
+        // This modal is used for topic editing only.
+        const select_stream_id = undefined;
+        dialog_widget.show_dialog_spinner();
+        message_edit.with_first_message_id(
+            current_stream_id,
+            old_topic_name,
+            (message_id) => {
+                message_edit.move_topic_containing_message_to_stream(
+                    message_id,
+                    select_stream_id,
+                    new_topic_name,
+                    send_notification_to_new_thread,
+                    send_notification_to_old_thread,
+                    "change_all",
+                );
+            },
+            (xhr) => {
+                dialog_widget.hide_dialog_spinner();
+                ui_report.error(
+                    $t_html({defaultMessage: "Error renaming topic"}),
+                    xhr,
+                    $("#rename_topic_modal #dialog_error"),
+                );
+            },
+        );
+    }
+
+    function rename_topic_post_render() {
+        const $topic_input = $("#rename_topic_form .inline_topic_edit");
+        composebox_typeahead.initialize_topic_edit_typeahead(
+            $topic_input,
+            current_stream_name,
+            false,
+        );
+        $("#rename_topic_modal .dialog_submit_button").prop("disabled", true);
+        $topic_input.on("input", () => {
+            // Pass current_stream_id as second parameter, since we do not allow
+            // changing stream in this modal.
+            update_submit_button_disabled_state($("#rename_topic_modal"), current_stream_id);
+        });
+    }
+
+    dialog_widget.launch({
+        html_heading: $t_html({defaultMessage: "Rename topic"}),
+        html_body: render_rename_topic_modal(args),
+        html_submit_button: $t_html({defaultMessage: "Confirm"}),
+        id: "rename_topic_modal",
+        on_click: rename_topic,
+        loading_spinner: true,
+        post_render: rename_topic_post_render,
+    });
+}
```

Comment:
It would be nice if we could avoid duplicating so much code from the "move topic" popover; in theory the two variants could differ only in the HTML template (strings and what inputs are displayed)? Idunno, may not be worth a lot of effort, but at least splitting these to their own file should help make it easier to avoid accidentally making changes in only one or the other variant.